consensus/hotstuff/pacemaker/pacemaker.go

```diff
@@ -224,7 +160,66 @@ func (p *ActivePaceMaker) Start(ctx context.Context) {
 	p.notifier.OnStartingTimeout(timerInfo)
 }
 
-// BlockRateDelay returns the delay for broadcasting its own proposals.
-func (p *ActivePaceMaker) BlockRateDelay() time.Duration {
-	return p.timeoutControl.BlockRateDelay()
+/* ------------------------------------ recovery parameters for PaceMaker ------------------------------------ */
+
+// recoveryInformation provides optional information to the PaceMaker during its construction
+// to ingest additional information that was potentially lost during a crash or reboot.
+// Following the "information-driven" approach, we consider potentially older or redundant
+// information as consistent with our already-present knowledge, i.e. as a no-op.
+type recoveryInformation func(p *ActivePaceMaker) error
+
+// WithQC informs the PaceMaker about the given QCs. Old and nil QCs are accepted (no-op).
+func WithQCs(qcs ...*flow.QuorumCertificate) recoveryInformation {
+	// To avoid excessive data base writes during initialization, we pre-filter the newest QC
+	// here and only hand that one to the viewTracker. For recovery, we allow the special case
+	// of nil QCs, because the genesis block has no QC.
+	var newestQC *flow.QuorumCertificate
+	for _, qc := range qcs {
+		if qc == nil {
+			continue // no-op
+		}
+		if newestQC == nil || newestQC.View < qc.View {
+			newestQC = qc
+		}
+	}
+	if newestQC == nil {
+		return func(p *ActivePaceMaker) error { return nil } // no-op
+	}
+
+	return func(p *ActivePaceMaker) error {
+		_, err := p.viewTracker.ProcessQC(newestQC) // panics for nil input
+		return err
+	}
+}
+
+// WithTC informs the PaceMaker about the given TCs. Old and nil TCs are accepted (no-op).
+func WithTCs(tcs ...*flow.TimeoutCertificate) recoveryInformation {
+	var newestTC *flow.TimeoutCertificate
+	var newestQC *flow.QuorumCertificate
+	for _, tc := range tcs {
```

Comment:
same here, can avoid many checks if used with QC and TC trackers

Reply:
👍 👉  https://github.com/onflow/flow-go/pull/4155/commits/ddcd4a65aedcc49bdd608ae8bf513161107f6a61